bentoml/_internal/resource.py

```diff
@@ -0,0 +1,279 @@
+from __future__ import annotations
+
+import os
+import re
+import math
+import typing as t
+import logging
+import functools
+from abc import ABCMeta
+from abc import abstractmethod
+
+import psutil
+
+from ..exceptions import BentoMLException
+
+logger = logging.getLogger(__name__)
+
+_RESOURCE_REGISTRY: dict[str, t.Type[Resource[t.Any]]] = {}
+
+T = t.TypeVar("T")
+
+
+def get_resource(resources: dict[str, t.Any], resource_kind: str) -> t.Any:
+    if resource_kind not in _RESOURCE_REGISTRY:
+        raise BentoMLException(f"Unknown resource kind '{resource_kind}'.")
+
+    resource: t.Type[Resource[t.Any]] = _RESOURCE_REGISTRY[resource_kind]
+
+    if resource_kind in resources:
+        if resources[resource_kind] == "system":
+            return resource.from_system()
+        else:
+            return resource.from_spec(resources[resource_kind])
+    else:
+        return None
+
+
+def system_resources() -> dict[str, t.Any]:
+    res: dict[str, t.Any] = {}
+    for resource_kind, resource in _RESOURCE_REGISTRY.items():
+        res[resource_kind] = resource.from_system()
+    return res
+
+
+class ResourceMeta(ABCMeta):
+    def __new__(  # pylint: disable=arguments-differ # we don't want to be generic
+        cls,
+        name: str,
+        bases: tuple[type],
+        clsdict: dict[t.Any, t.Any],
+        *,
+        resource_id: str,
+    ):
+        res = super().__new__(cls, name, bases, clsdict)
```

Comment:
very very nit thing
```suggestion
        dct: dict[t.Any, t.Any],
        *,
        resource_id: str,
    ):
        res = super().__new__(cls, name, bases, dct)
```

Reply:
Thought about this one a bit and I think `attrs` is probably better than both.

Reply:
Named it `attr_dict` instead; I think that's decent?

Reply:
I think it is a convention for writting __new__ to use dct right?

Reply:
AFAICT there is no convention here. `ABCMeta`'s `__new__` uses `namespace` (which is confusing, I think), `type`'s stubs use `__dict` (which is pretty bad). I've only seen `dct` used in one tutorial, I think.